core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -627,19 +626,36 @@ public PlanOptimizers(
                 .addAll(projectionPushdownRules)
                 .add(new PushProjectionIntoTableScan(metadata, typeAnalyzer, scalarStatsCalculator))
                 .add(new RemoveRedundantIdentityProjections())
+                .add(new PushLimitThroughMarkDistinct())
```

Comment:
add a comment why this is inside pushIntoTableScanRulesExceptJoins set, otherwise looks like unintentional

maybe it shouldn't be here? instead, `PushLimitThroughMarkDistinct` could come together with MultipleDistinctAggregationToMarkDistinct?

Reply:
>maybe it shouldn't be here? instead, PushLimitThroughMarkDistinct could come together with MultipleDistinctAggregationToMarkDistinct?

per @kasiafi 
> After removing MultipleDistinctAggregationToMarkDistinct from an IterativeOptimizer, there's no way that a MarkDistinctNode could be present at that stage, so the rule PushLimitThroughMarkDistinct can be also removed. However, I'd keep it there next to other PushLimit rules in case it's needed again in the future. Actually, the PushLimit rules could be extracted like projectionPushdownRules.

Reply:
never mind. Reworking after chat with @kasiafi